### ISSUE TITLE
Performance vs Batching Promise.all

### ISSUE THREAD

**Issue by Jack-Byrne** [supercharge/promise-pool#70]
Hello I was looking to implement this package for large sets of db reads and writes. I was comparing using the PromisePool with concurrency of 1000 to using Promise.all in batches of 1000.

Performance testing was done on 100k db read and write operations.
Using promise pools the execution time for 100k total operations with a concurrency of 1000 was `1:40.937 (m:ss.mmm)`
Using 1000 batch size for Promise.all(...) had an execution time of `41.867s` (also on 100k db reads and writes)

I tested multiple times on different sets of data to ensure there weren't some random anomalies and to avoid performance variability from db caching.

I figured promise pools would be significantly faster since it should be maintaining 1000 open promises at a time where the batch Promse.all approach needs to wait for all 1000 operations in a batch to complete before starting the next 1000 operations.

Is this expected and are there ways to optimize the use of this library?

Node.js: v18.12.1

```js
  const batchSize = 1000;
  const entries=[..];
  const findAndInsert = async (number) => {
    const item = await getItem(number);
    await itemMapper.insert({ ...item });
  };
  
  console.time("promises"); //start timer
  
  await PromisePool.for(entries)
    .withConcurrency(batchSize)
    .process(findAndInsert);
    
  console.timeEnd("promises"); // end timer
```

**Comment by marcuspoehls** [comment]
@Jack-Byrne Hey Jack, thank you for your performance testing. My first impression is also that the promise pool shouldn’t have this large overhead in terms of processing time. Interesting results. Can you share your benchmarking code? 

I would love to improve the performance of this package. Please feel free to suggest improvements or send a pull request. I appreciate any type of help 😃

Thank you for sharing your ideas and results!

**Comment by WillianAgostini** [comment]
Hi @marcuspoehls 

I was able to simulate a scenario where there is a divergence in time between PromisePool and Promise.all, but with very high values of promises, so it must be evaluated whether it is a possible scenario.

The gist below contains the code that has this simulation, where it obtained the following response times:
https://gist.github.com/WillianAgostini/7e18fc25cc14a7f9c8676a2572ed76bf

for 50 000 promises, got:
- promises: 3037s
- promisesAll: 1,009s

for 10 000 promises, got:
- PromisePool: 1,069s
- Promise.all: 1,000s

Any number of promises below that the difference is negligible.

**Comment by marcuspoehls** [comment]
@WillianAgostini Hey Willian, thank you for sharing your benchmark. That’s helpful 👍

**Comment by marcuspoehls** [comment]
This was fixed in #71 which landed in `main` today. The promise pool and `Promise.all` are now close together, even for 50k promises
